### PR TITLE
Refactor workflow to extract processing with submission

### PR DESCRIPTION
**Pull request by bdc34**
This is the direction I was going with the change to workflow. 

It is mainly moving the workfow into a WorkfowDefinition and having a separate WorkflowProcessor that handles where a submission is in a workflow.

This is a early preview PR. After this I'd test the workflows and integrate with routes and flow_control.py.

ARXIVNG-2886

### CODE REVIEW

File: submit/workflow/conditions.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+from arxiv.submission.domain import Submission
```
Comment (bdc34) [inline submit/workflow/conditions.py]: I split the conditions on the Workflow stages out for reuse and testing.

File: submit/workflow/conditions.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    
+
+def source_processed(submission: Submission) -> bool:
+    """Determine whether the submitter has compiled their upload."""
+    non_processing_content=(
+        submission.source_content.source_format != SubmissionContent.Format.TEX
+        and
+        submission.source_content.source_format != SubmissionContent.Format.POSTSCRIPT)
+    return has_valid_content(submission) and \
```
Comment (bdc34) [inline submit/workflow/conditions.py:49]: Here is the reuse. Not as much as I was hoping.

File: submit/workflow/processor.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+"""Defines submission stages and workflows supported by this UI."""
```
Comment (bdc34) [inline submit/workflow/processor.py:1]: This is a new Class that handles where a submission is within a WorkflowDefinition.

File: submit/workflow/test_new_submission.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class TestNewSubmissionWorkflow(TestCase):
+
+    def testVerifyUser(self):
+        seen = {}
+        submission = mock.MagicMock(submission_id=1234,
+                                    is_finalized=False,
+                                    submitter_contact_verified=False)        
+        nswfps = processor.WorkflowProcessor(
```
Comment (bdc34) [inline submit/workflow/test_new_submission.py]: This shows how the workflow is used.

File: submit/workflow/conditions.py (hunk 1 above)
Comment (DavidLFielding) [reply submit/workflow/conditions.py]: Seems reasonable.

File: submit/workflow/test_new_submission.py (hunk 4 above)
Comment (DavidLFielding) [reply submit/workflow/test_new_submission.py]: Test is helpful. Hopefully we'll have more complicated tests in future though this depends on what gets decided in terms of locking down workflow. I'm more in favor of allowing user to select order of steps they complete while disabling step transitions that are invalid (ie., process without uploading files)

**Review by DavidLFielding — APPROVED** [review]
Looks fine to me. Mostly reorganization/refactoring at this point. I'm not clear on expected reuse of conditionals. I can't wait until this is wired up so we can move forward tightening up workflow and eliminating invalid transitions.

File: submit/workflow/conditions.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
+from arxiv.submission.domain import Submission
+
+def contact_verified(submission: Submission) -> bool:
```
Comment (mhl10) [inline submit/workflow/conditions.py]:
```suggestion
def is_contact_verified(submission: Submission) -> bool:
```

File: submit/workflow/conditions.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    """Determine whether the submitter has verified their information."""
+    return submission.submitter_contact_verified is True
+
+
+def authorship_indicated(submission: Submission) -> bool:
```
Comment (mhl10) [inline submit/workflow/conditions.py]:
```suggestion
def is_authorship_indicated(submission: Submission) -> bool:
```

File: submit/workflow/conditions.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    return submission.submitter_is_author is not None
+
+
+def has_license(submission: Submission) -> bool:
+    """Determine whether the submitter has selected a license."""
+    return submission.license is not None
+
+
+
+def policy_accepted(submission: Submission) -> bool:
```
Comment (mhl10) [inline submit/workflow/conditions.py]:
```suggestion
def is_policy_accepted(submission: Submission) -> bool:
```

File: submit/workflow/conditions.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+def has_valid_content(submission: Submission) -> bool:
+    """Determine whether the submitter has uploaded files."""
+    return  submission.source_content is not None and\
+        submission.source_content.checksum is not None and\
+        submission.source_content.source_format != SubmissionContent.Format.INVALID
+
+    
+
+def source_processed(submission: Submission) -> bool:
```
Comment (mhl10) [inline submit/workflow/conditions.py]:
```suggestion
def is_source_processed(submission: Submission) -> bool:
```

File: submit/workflow/conditions.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    """Determine whether the submitter has compiled their upload."""
+    non_processing_content=(
+        submission.source_content.source_format != SubmissionContent.Format.TEX
+        and
+        submission.source_content.source_format != SubmissionContent.Format.POSTSCRIPT)
+    return has_valid_content(submission) and \
+            (submission.is_source_processed or non_processing_content)
+
+
+def metadata_complete(submission: Submission) -> bool:
```
Comment (mhl10) [inline submit/workflow/conditions.py]:
```suggestion
def is_metadata_complete(submission: Submission) -> bool:
```

File: submit/workflow/conditions.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    """Determine whether the submitter has entered required metadata."""
+    return (submission.metadata.title is not None
+            and submission.metadata.abstract is not None
+            and submission.metadata.authors_display is not None)
+
+def opt_metadata_complete(submission: Submission) -> bool:
```
Comment (mhl10) [inline submit/workflow/conditions.py]:
```suggestion
def is_opt_metadata_complete(submission: Submission) -> bool:
```

File: submit/workflow/__init__.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
+"""Defines submission stages and workflows supported by this UI."""
+
+from typing import Iterable, Optional, Callable, List, Iterator
+
+from arxiv.submission.domain import Submission
+from dataclasses import dataclass
+
+from . import stages
+from .stages import Stage
```
Comment (mhl10) [inline submit/workflow/__init__.py:9]: absolute imports may be preferable for clarity

File: submit/workflow/processor.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+from . import WorkflowDefinition, Stage
+
+
+@dataclass
+class WorkflowProcessor:
+    workflow: WorkflowDefinition
+    submission: Submission
+    seen: Dict[str, bool] = field(default_factory=dict)
+
+    def complete(self) -> bool:
```
Comment (mhl10) [inline submit/workflow/processor.py]:
```suggestion
    def is_complete(self) -> bool:
```

File: submit/workflow/processor.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+            or all(map(self.is_done, self.workflow.iter_prior(self.workflow, stage)))
+
+    def current_stage(self) -> Optional[Stage]:
+        """Get the first stage in the workflow that is not done."""
+        for stage in self.workflow.order:
+            if not self.is_done(stage):
+                return stage
+        return None
+
+    def is_complete(self, stage: Optional[Stage]) -> bool:
```
Comment (mhl10) [inline submit/workflow/processor.py]:
```suggestion
    def is_stage_complete(self, stage: Optional[Stage]) -> bool:
```

File: submit/workflow/__init__.py (hunk 11 above)
Comment (bdc34) [reply submit/workflow/__init__.py:9]: I was thinking that this was a place  where the stages are closely related to the workflows.

**Comment by bdc34** [comment]
Thanks for you feedback. I've incorporated some of your ideas. 

I'm going to close this PR and open a new one once I have it wired into the flask app.

**Pull request closed without merging**